### ISSUE TITLE
bug: ALZ Vending - issue with Online MG - Error: could not read properties for Management Group "xxx-online"

### ISSUE THREAD

**Issue by cbezenco** [Azure/terraform-azurerm-lz-vending#238]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Versions

Please paste the output of `terraform version` command from within the initialized directory:

```text
Terraform v1.5.4
on windows_amd64
+ provider registry.terraform.io/azure/azapi v1.8.0
+ provider registry.terraform.io/hashicorp/azurerm v3.68.0
```

Please enter the module version that you are using:

```text
module "caf-enterprise-scale" {
  source  = "Azure/caf-enterprise-scale/azurerm"
  version = "4.2.0"
module "lz_vending" {
  source   = "Azure/lz-vending/azurerm"
  version  = "=3.4.1" 
```

#### Steps to Reproduce

1. step - Start from scratch - Terraform Destroy ALZ - moving back all spoke subscription in Default management group for new subscriptions
2. step - Deploy ALZ (in my env I am using hub and spoke with 1-4 regional hub depending on config)
3. step - Deploy ALZ vending - my spoke are targeting the \ALZ\Landing Zones\Online MG

#### Screenshots
Output extract from terraform plan (in this example, I am provisioning 2 subscriptions). 

Plan: 10 to add, 4 to change, 2 to destroy.
╷
│ Error: could not read properties for Management Group "dev-online"
│
│   with module.lz_vending["LDE-DEV-LZ02"].module.subscription[0].azurerm_management_group_subscription_association.this[0],
│   on .terraform\modules\lz_vending\modules\subscription\main.tf line 13, in resource "azurerm_management_group_subscription_association" "this":
│   13: resource "azurerm_management_group_subscription_association" "this" {
│
╵
╷
│ Error: could not read properties for Management Group "dev-online"
│
│   with module.lz_vending["LDE-DEV-LZ01"].module.subscription[0].azurerm_management_group_subscription_association.this[0],
│   on .terraform\modules\lz_vending\modules\subscription\main.tf line 13, in resource "azurerm_management_group_subscription_association" "this":
│   13: resource "azurerm_management_group_subscription_association" "this" {
│
╵
#### Additional context

- No issue if I target \ALZ\Landing Zones
- Moving manually one subscription from the Default management group for new subscriptions into \ALZ\Landing Zones\Online MG solve the issue. I can Terraform apply to get the remaining subscription under \ALZ\Landing Zones\Online MG.

**Comment by cbezenco** [comment]
Interestingly, if I run terraform destroy after a successful deployment (using manual workaround described above), I am getting the following error message 

PS C:\github\tf-eslz\alzVending> terraform destroy
module.lz_vending["LDE-DEV-LZ02"].module.subscription[0].azurerm_management_group_subscription_association.this[0]: Refreshing state... [id=/managementGroup/dev-online/subscription/b454d5bb-af38-4b00-9377-579229cd0b85]
module.lz_vending["LDE-DEV-LZ01"].module.subscription[0].azurerm_management_group_subscription_association.this[0]: Refreshing state... [id=/managementGroup/dev-online/subscription/c7cadae2-f210-4710-8b9e-e4022c8be0c9]
module.lz_vending["LDE-DEV-LZ02"].azapi_resource.telemetry_root[0]: Refreshing state... [id=/subscriptions/b454d5bb-af38-4b00-9377-579229cd0b85/providers/Microsoft.Resources/deployments/pid-50a8a460-d517-4b11-b86c-6de447806b67_2.1.0_00010302]
module.lz_vending["LDE-DEV-LZ01"].azapi_resource.telemetry_root[0]: Refreshing state... [id=/subscriptions/c7cadae2-f210-4710-8b9e-e4022c8be0c9/providers/Microsoft.Resources/deployments/pid-50a8a460-d517-4b11-b86c-6de447806b67_2.1.0_00010302]
module.lz_vending["LDE-DEV-LZ01"].module.virtualnetwork[0].azapi_resource.rg["rg-vnet-dev-lz01-01"]: Refreshing state... [id=/subscriptions/c7cadae2-f210-4710-8b9e-e4022c8be0c9/resourceGroups/rg-vnet-dev-lz01-01]
module.lz_vending["LDE-DEV-LZ02"].module.virtualnetwork[0].azapi_resource.rg["rg-vnet-dev-lz02-03"]: Refreshing state... [id=/subscriptions/b454d5bb-af38-4b00-9377-579229cd0b85/resourceGroups/rg-vnet-dev-lz02-03]
module.lz_vending["LDE-DEV-LZ02"].module.virtualnetwork[0].azapi_resource.rg["rg-vnet-dev-lz02-02"]: Refreshing state... [id=/subscriptions/b454d5bb-af38-4b00-9377-579229cd0b85/resourceGroups/rg-vnet-dev-lz02-02]
module.lz_vending["LDE-DEV-LZ02"].module.virtualnetwork[0].azapi_resource.rg["rg-vnet-dev-lz02-01"]: Refreshing state... [id=/subscriptions/b454d5bb-af38-4b00-9377-579229cd0b85/resourceGroups/rg-vnet-dev-lz02-01]
module.lz_vending["LDE-DEV-LZ02"].module.virtualnetwork[0].azapi_resource.vnet["tertiary"]: Refreshing state... [id=/subscriptions/b454d5bb-af38-4b00-9377-579229cd0b85/resourceGroups/rg-vnet-dev-lz02-03/providers/Microsoft.Network/virtualNetworks/vnet-dev-lz02-03]
module.lz_vending["LDE-DEV-LZ02"].module.virtualnetwork[0].azapi_resource.vnet["secondary"]: Refreshing state... [id=/subscriptions/b454d5bb-af38-4b00-9377-579229cd0b85/resourceGroups/rg-vnet-dev-lz02-02/providers/Microsoft.Network/virtualNetworks/vnet-dev-lz02-02]
module.lz_vending["LDE-DEV-LZ01"].module.virtualnetwork[0].azapi_resource.vnet["primary"]: Refreshing state... [id=/subscriptions/c7cadae2-f210-4710-8b9e-e4022c8be0c9/resourceGroups/rg-vnet-dev-lz01-01/providers/Microsoft.Network/virtualNetworks/vnet-dev-lz01-01]
module.lz_vending["LDE-DEV-LZ02"].module.virtualnetwork[0].azapi_resource.vnet["primary"]: Refreshing state... [id=/subscriptions/b454d5bb-af38-4b00-9377-579229cd0b85/resourceGroups/rg-vnet-dev-lz02-01/providers/Microsoft.Network/virtualNetworks/vnet-dev-lz02-01]
module.lz_vending["LDE-DEV-LZ02"].module.virtualnetwork[0].azapi_update_resource.vnet["secondary"]: Refreshing state... [id=/subscriptions/b454d5bb-af38-4b00-9377-579229cd0b85/resourceGroups/rg-vnet-dev-lz02-02/providers/Microsoft.Network/virtualNetworks/vnet-dev-lz02-02]
module.lz_vending["LDE-DEV-LZ02"].module.virtualnetwork[0].azapi_update_resource.vnet["primary"]: Refreshing state... [id=/subscriptions/b454d5bb-af38-4b00-9377-579229cd0b85/resourceGroups/rg-vnet-dev-lz02-01/providers/Microsoft.Network/virtualNetworks/vnet-dev-lz02-01]
module.lz_vending["LDE-DEV-LZ02"].module.virtualnetwork[0].azapi_update_resource.vnet["tertiary"]: Refreshing state... [id=/subscriptions/b454d5bb-af38-4b00-9377-579229cd0b85/resourceGroups/rg-vnet-dev-lz02-03/providers/Microsoft.Network/virtualNetworks/vnet-dev-lz02-03]
module.lz_vending["LDE-DEV-LZ01"].module.virtualnetwork[0].azapi_update_resource.vnet["primary"]: Refreshing state... [id=/subscriptions/c7cadae2-f210-4710-8b9e-e4022c8be0c9/resourceGroups/rg-vnet-dev-lz01-01/providers/Microsoft.Network/virtualNetworks/vnet-dev-lz01-01]
module.lz_vending["LDE-DEV-LZ02"].module.virtualnetwork[0].azapi_resource.peering_hub_inbound["tertiary"]: Refreshing state... [id=/subscriptions/2560ed2d-c2de-48d7-b953-cb112107a0f5/resourceGroups/dev-connectivity-northeurope/providers/Microsoft.Network/virtualNetworks/dev-hub-northeurope/virtualNetworkPeerings/peer-c7d82284-0c50-5951-a907-c4f281cdf891]
module.lz_vending["LDE-DEV-LZ02"].module.virtualnetwork[0].azapi_resource.peering_hub_inbound["secondary"]: Refreshing state... [id=/subscriptions/2560ed2d-c2de-48d7-b953-cb112107a0f5/resourceGroups/dev-connectivity-northeurope/providers/Microsoft.Network/virtualNetworks/dev-hub-northeurope/virtualNetworkPeerings/peer-d8f3937c-ec78-5de3-9b3c-5518df8216a6]
module.lz_vending["LDE-DEV-LZ01"].module.virtualnetwork[0].azapi_resource.peering_hub_inbound["primary"]: Refreshing state... [id=/subscriptions/2560ed2d-c2de-48d7-b953-cb112107a0f5/resourceGroups/dev-connectivity-northeurope/providers/Microsoft.Network/virtualNetworks/dev-hub-northeurope/virtualNetworkPeerings/peer-a96a2006-569a-5088-91dc-f8c6099839cf]
module.lz_vending["LDE-DEV-LZ02"].module.virtualnetwork[0].azapi_resource.peering_hub_inbound["primary"]: Refreshing state... [id=/subscriptions/2560ed2d-c2de-48d7-b953-cb112107a0f5/resourceGroups/dev-connectivity-northeurope/providers/Microsoft.Network/virtualNetworks/dev-hub-northeurope/virtualNetworkPeerings/peer-494ff583-a062-5c80-a96a-27c95d90df86]
module.lz_vending["LDE-DEV-LZ02"].module.virtualnetwork[0].azapi_resource.peering_hub_outbound["tertiary"]: Refreshing state... [id=/subscriptions/b454d5bb-af38-4b00-9377-579229cd0b85/resourceGroups/rg-vnet-dev-lz02-03/providers/Microsoft.Network/virtualNetworks/vnet-dev-lz02-03/virtualNetworkPeerings/peer-99063ec4-d195-51f4-b0ab-54b76e38ce24]
module.lz_vending["LDE-DEV-LZ02"].module.virtualnetwork[0].azapi_resource.peering_hub_outbound["secondary"]: Refreshing state... [id=/subscriptions/b454d5bb-af38-4b00-9377-579229cd0b85/resourceGroups/rg-vnet-dev-lz02-02/providers/Microsoft.Network/virtualNetworks/vnet-dev-lz02-02/virtualNetworkPeerings/peer-99063ec4-d195-51f4-b0ab-54b76e38ce24]
module.lz_vending["LDE-DEV-LZ01"].module.virtualnetwork[0].azapi_resource.peering_hub_outbound["primary"]: Refreshing state... [id=/subscriptions/c7cadae2-f210-4710-8b9e-e4022c8be0c9/resourceGroups/rg-vnet-dev-lz01-01/providers/Microsoft.Network/virtualNetworks/vnet-dev-lz01-01/virtualNetworkPeerings/peer-99063ec4-d195-51f4-b0ab-54b76e38ce24]
module.lz_vending["LDE-DEV-LZ02"].module.virtualnetwork[0].azapi_resource.peering_hub_outbound["primary"]: Refreshing state... [id=/subscriptions/b454d5bb-af38-4b00-9377-579229cd0b85/resourceGroups/rg-vnet-dev-lz02-01/providers/Microsoft.Network/virtualNetworks/vnet-dev-lz02-01/virtualNetworkPeerings/peer-99063ec4-d195-51f4-b0ab-54b76e38ce24]
╷
│ Error: could not read properties for Management Group "dev-online"
│
│   with module.lz_vending["LDE-DEV-LZ02"].module.subscription[0].azurerm_management_group_subscription_association.this[0],
│   on .terraform\modules\lz_vending\modules\subscription\main.tf line 13, in resource "azurerm_management_group_subscription_association" "this":
│   13: resource "azurerm_management_group_subscription_association" "this" {
│
╵
╷
│ Error: could not read properties for Management Group "dev-online"
│
│   with module.lz_vending["LDE-DEV-LZ01"].module.subscription[0].azurerm_management_group_subscription_association.this[0],
│   on .terraform\modules\lz_vending\modules\subscription\main.tf line 13, in resource "azurerm_management_group_subscription_association" "this":
│   13: resource "azurerm_management_group_subscription_association" "this" {

**Comment by jtracey93** [comment]
Can you share a gist or code snippet of your LZ vending and ALZ configs here @cbezenco ?

**Comment by cbezenco** [comment]
Here is the code

locals {
  # spokes_data_dir is the directory containing the YAML files for the landing zones.
  spokes_data_dir = "../maps/spokes.yaml"

  # spokes_files is the list of landing zone YAML files to be processed
  spokes_files = fileset(local.spokes_data_dir, "${var.instance}*")

  # landing_zone_data_map is the decoded YAML data stored in a map
  spokes_data_map = {
    for f in local.spokes_files :
    f => yamldecode(file("${local.spokes_data_dir}/${f}"))
  }
}

# The landing zone module will be called once per spokes_*.yaml file
# in the data directory.
module "lz_vending" {
  source   = "Azure/lz-vending/azurerm"
  version  = "=3.4.1" # change this to your desired version, https://www.terraform.io/language/expressions/version-constraints
  for_each = local.spokes_data_map

  location = each.value.location

  # subscription variables
  subscription_id = each.value.subscription_id

  # management group association variables
  subscription_management_group_association_enabled = true
  subscription_management_group_id                  = each.value.management_group_id

  # virtual network variables
  virtual_network_enabled = true
  virtual_networks        = each.value.virtual_networks

  # role assignment variables
  role_assignment_enabled = true
  role_assignments        = each.value.role_assignments

}

**Comment by cbezenco** [comment]
Here is one of the sub config yaml file
```
---
name: LDE-DEV-NE-LZ01
workload: dev
location: northeurope
subscription_id: 123456-f210-4710-8b9e-e4022c8be0c9
management_group_id: dev-online
virtual_networks:
  primary:
    name: vnet-dev-lz01-01
    location : northeurope
    resource_group_name: rg-vnet-dev-lz01-01
    resource_group_lock_enabled: false
    address_space:
      - "10.51.0.0/16"
    hub_peering_enabled: true
    hub_network_resource_id: /subscriptions/123456-c2de-48d7-b953-cb112107a0f5/resourceGroups/dev-connectivity-northeurope/providers/Microsoft.Network/virtualNetworks/dev-hub-northeurope
role_assignments:
  owner_to_user:
    principal_id: 123456-c8d7-409a-97ce-d01a5215a038
    definition: Owner
    relative_scope: ''
  contrib_to_group:
    principal_id: a95e6a31-6627-46f5-862c-f499ddffe88e
    definition: Contributor
    relative_scope: ''
  ```

**Comment by jtracey93** [comment]
@cbezenco what about the ALZ module config?

**Comment by cbezenco** [comment]
Here is the ALZ module config.

```
module "caf-enterprise-scale" {
  source  = "Azure/caf-enterprise-scale/azurerm"
  version = "4.2.0"

  providers = {
    azurerm              = azurerm
    azurerm.management   = azurerm.management
    azurerm.connectivity = azurerm.connectivity
  }

  # set TF_VAR_env variable to "dev" to develop in interactive mode
  root_parent_id   = data.azurerm_client_config.core.tenant_id
  root_id          = var.env
  root_name        = "ALZ-${var.env}"
  default_location = var.default_location
  library_path     = "${path.root}/lib.${var.env}"

  deploy_demo_landing_zones = false

  custom_landing_zones = {
    "${var.env}-online-example-1" = {
      display_name               = "${upper(var.env)} Online Example 1"
      parent_management_group_id = "${var.env}-landing-zones"
      subscription_ids           = []
      archetype_config = {
        archetype_id   = "customer_online"
        parameters     = {}
        access_control = {}
      }
    }
    "${var.env}-online-example-2" = {
      display_name               = "${upper(var.env)} Online Example 2"
      parent_management_group_id = "${var.env}-landing-zones"
      subscription_ids           = []
      archetype_config = {
        archetype_id = "customer_online"
        parameters = {
          Deny-Resource-Locations = {
            listOfAllowedLocations = ["westeurope", "northeurope", "eastus", "westus"]
          }
          Deny-RSG-Locations = {
            listOfAllowedLocations = ["westeurope", "northeurope", "eastus", "westus"], 
          }
        }
        access_control = {}
      }
    }
  }


  # Enable deployment of the management resources, using the management
  # aliased provider to populate the correct Subscription ID
  deploy_management_resources    = true
  subscription_id_management     = data.azurerm_client_config.management.subscription_id
  configure_management_resources = local.configure_management_resources

  # Enable deployment of the connectivity resources, using the connectivity
  # aliased provider to populate the correct Subscription ID
  deploy_connectivity_resources    = true
  subscription_id_connectivity     = data.azurerm_client_config.connectivity.subscription_id
  configure_connectivity_resources = local.configure_connectivity_resources
  
  deploy_online_landing_zones = true

}

**Comment by jtracey93** [comment]
Thanks. I might be going mad, but i don't see a MG declared called `dev-online` in your ALZ code?

**Comment by jtracey93** [comment]
Okay so `var.env` == `dev1, cool

This may be because of module dependencies, have you got a dependency on the vending modules against the ALZ module?

E.g. the MG need to exist before you can deploy and move a sub into it

**Comment by cbezenco** [comment]
I don't believe I have any (not sure I understood what kind of dependancy you think about). 
I keep alz and alzvending into separate deployment (folder/state). In my dev lab, I trigger those 2 terraform deployment manually / sequentially (no DevOps pipeline). 
Before deploying the alzvending, I can see the Online MG is deployed.  
What is strange is before I have been using the Landing Zone MG without issue. As one of the workload that I used required Internet access and was blocked by standard ALZ policies, I switched to Online MG and this is where i started to see the issue. 
It really look like the failure happen only if the Online MG is empty. As soon as I move one subscription (even it is not one of the subscription configured), I see no error.

**Comment by jtracey93** [comment]
ah okay i get you now, i thought this was a single TF plan and apply in a single directory.

Have you tested just new management groups, outside the ALZ module via TF and then trying to get the subs to go to that from lz vending?

**Comment by cbezenco** [comment]
Ok I tested with dev-online-example-1 and dev-online-example-2 and at the beginning it worked but then I moved manually the sub into the new subscription MG to do additional test with dev-online-example-1 I started to face the same error (alzvending state still expect the sub to be in dev-online-example-2) ... I think I understand now what is happening. 

To reduce cost in my lab, I decomissioned ALZ hub(s) every evening and for some reasons that I do not understand, decomissioning all the hubs has the side effect of moving the configured spoke into the management group configured "Default management group for new subscriptions". This side effect means that the alzvending state is now longuer valid (similar situation as the one described above with dev-online-example-2) and when I do a Terraform Apply, I get the error message reported.

Looks like to solve the issue, I will need to keep alz and alzvending module within the same state file. 

It might be interesting to understand why decomissioning all the hubs in ALZ triggers a move of the spoke into the  "Default management group for new subscriptions".
